internal/tescale/scale.go

```diff
@@ -0,0 +1,267 @@
+	tunnelNHGRatio := param.V4TunnelCount / param.V4TunnelNHGCount
+	for idx, ip := range v4TunnelIPAddrs.AllIPs() {
+		if idx%tunnelNHGRatio == 0 {
+			nhgID = idPool.NextNHGID()
+
+			// Build NHs and link NHs to NHG.
+			for i := 0; i < param.V4TunnelNHGSplitCount; i++ {
+				vip := v4VIPAddrs.NextIP()
+				nhID = idPool.NextNHID()
+				vrfDefault.NHs = append(vrfDefault.NHs,
+					fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithIPAddress(vip),
+				)
+				vrfDefault.NHGs = append(vrfDefault.NHGs,
+					fluent.NextHopGroupEntry().WithID(nhgID).AddNextHop(nhID, 1).WithNetworkInstance(defaultVRF).WithBackupNHG(nhgRedirectToVrfR),
+				)
+			}
+		}
+
+		// Build IPv4 entry
+		vrfTConf.V4Entries = append(vrfTConf.V4Entries,
+			fluent.IPv4Entry().WithPrefix(ip).WithNextHopGroup(nhgID).WithNetworkInstance(VRFT).WithNextHopGroupNetworkInstance(defaultVRF),
+		)
+	}
+
+	// Default VRF:
+
+	// * each VIP 1:1 map to a NHG
+	// * each NHG points to unique NHs
+	for _, ip := range v4VIPAddrs.AllIPs() {
+		nhgID := idPool.NextNHGID()
+
+		// Build NHs and link NHs to NHG.
+		for i := 0; i < param.EgressNHGSplitCount; i++ {
+			vip := v4EgressIPAddrs.AllIPs()[i]
+			nhID = idPool.NextNHID()
+			vrfDefault.NHs = append(vrfDefault.NHs,
+				fluent.NextHopEntry().WithIndex(nhID).WithNetworkInstance(defaultVRF).WithIPAddress(vip),
+			)
+			vrfDefault.NHGs = append(vrfDefault.NHGs,
+				fluent.NextHopGroupEntry().WithID(nhgID).AddNextHop(nhID, 1).WithNetworkInstance(defaultVRF),
+			)
+		}
+
+		// Build IPv4 entry
+		vrfDefault.V4Entries = append(vrfDefault.V4Entries,
+			fluent.IPv4Entry().WithPrefix(ip).WithNextHopGroup(nhgID).WithNetworkInstance(defaultVRF).WithNextHopGroupNetworkInstance(defaultVRF),
+		)
+	}
+
+	// VRF_R
+
+	// build backup NHG and NH.
+	nhID = idPool.NextNHID()
+	nhgID = idPool.NextNHGID()
+	nhgDecapToDefault := nhgID
+	vrfDefault.NHs = append(vrfDefault.NHs,
+		fluent.NextHopEntry().WithIndex(nhID).WithDecapsulateHeader(fluent.IPinIP).WithNetworkInstance(defaultVRF).WithNextHopNetworkInstance(defaultVRF),
+	)
+	vrfDefault.NHGs = append(vrfDefault.NHGs,
+		fluent.NextHopGroupEntry().WithID(nhgID).AddNextHop(nhID, 1).WithNetworkInstance(defaultVRF),
+	)
+
+	// build IP entries and related NHG and NHs.
+	// * Each NHG 1:1 mapping to NH
+	// * Each NH has one entry for decap and encap
+	// * All NHG has a backup for decap then goto default VRF.
+	reEncapNHGRatio := param.V4TunnelCount / param.V4ReEncapNHGCount
+	for idx, ip := range v4TunnelIPAddrs.AllIPs() {
+		if idx%reEncapNHGRatio == 0 {
+			nhgID = idPool.NextNHGID()
+		}
+
+		nhID = idPool.NextNHID()
+		vrfDefault.NHs = append(vrfDefault.NHs,
+			fluent.NextHopEntry().WithIndex(nhID).WithDecapsulateHeader(fluent.IPinIP).WithEncapsulateHeader(fluent.IPinIP).
+				WithNetworkInstance(defaultVRF).WithIPinIP(tunnelSrcIP, v4TunnelIPAddrs.AllIPs()[(idx+1)%len(v4TunnelIPAddrs.AllIPs())]),
+		)
+		vrfDefault.NHGs = append(vrfDefault.NHGs,
```

Comment:
add to the list only when a new NHG id is allocated and created. This causing multiple ADD requests for the same NHG to the server